### PR TITLE
fix: preserve map frame through resize

### LINKED ISSUE #2172: [Bug]: Map flashes white while resizing the window
### Before you start

- [x] I searched the [existing issues](https://github.com/opengeos/GeoLibre/issues) and did not find a duplicate.
- [x] This is a bug report, not a usage question (questions belong in [Discussions](https://github.com/opengeos/GeoLibre/discussions)).

### What happened?

The map briefly flashes white while the browser or desktop window is being resized. Frame-level testing showed the WebGL framebuffer becoming transparent for roughly 100–150 ms, while the canvas and WebGL context remained mounted.

### Steps to reproduce

1. Open GeoLibre with the default globe basemap.
2. Continuously drag the window edge wider and narrower.
3. Observe the map area briefly flash white.

### Expected behavior

The map remains visible and visually stable throughout window resizing.

### How are you running GeoLibre?

Local web development build, GeoLibre 2.8.0 (d05f0903)

### Operating system and browser

macOS, Chromium

### Sample data

No sample data required; reproducible with the default globe basemap.

### Screenshots or logs

No resize-related console errors. Frame-level sampling showed the full WebGL framebuffer returning transparent pixels for three consecutive resize frames while the canvas and context remained active.

### PR DESCRIPTION
**Pull request by giswqs**
## Summary

- capture the current rendered map frame when a resize begins
- center the snapshot without changing its aspect ratio over a theme-matched background while the container changes
- keep it visible through the WebGL framebuffer clear, then remove it after MapLibre renders at the new size
- cover overlay cleanup and overlapping resize lifecycle edge cases in scheduler tests

## Verification

- reproduced the old canvas remaining 612 px wide while its container expanded to 1012 px
- confirmed the overlay covers the transparent framebuffer resize frame without deforming the globe
- verified the default globe through rapid Chromium viewport changes in light and dark themes
- `node --import tsx --test tests/map-resize.test.ts`
- `npm run build`
- `pre-commit run --files packages/map/src/map-resize.ts tests/map-resize.test.ts`

Fixes #2172

## Summary by CodeRabbit

* **Bug Fixes**
  * Map content now remains visually stable during window and panel resizing.
  * Added a temporary frame overlay to prevent flickering or blank areas while the map redraws.
  * The preserved view is automatically removed once rendering completes or resizing ends.
  * Improved handling for resize events where no movement occurs.

### CODE REVIEW

File: packages/map/src/map-resize.ts [hunk 1]
```diff
@@ -79,7 +116,11 @@ export function createMapResizeScheduler({
     resizeFrame = window.requestAnimationFrame(() => {
       resizeFrame = null;
       if (!mapNeedsResize()) return;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in c6bd68fc. No-op resize commits now remove the preserved frame, with regression coverage for a splitter click without movement.

File: packages/map/src/map-resize.ts [hunk 2]
```diff
@@ -79,7 +116,11 @@ export function createMapResizeScheduler({
...
-      getMap()?.resize();
+      const map = getMap();
+      if (!map) return;
+      preserveRenderedFrame();
+      map.once("render", removeFrameOverlay);
+      map.resize();
       observedDevicePixelRatio = window.devicePixelRatio;
     });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts:186]: Done in c6bd68fc. No-op resize commits now remove the preserved frame, with regression coverage for a splitter click without movement.

File: packages/map/src/map-resize.ts [hunk 3]
```diff
@@ -78,8 +115,18 @@ export function createMapResizeScheduler({
...
+        removeFrameOverlay();
+        return;
+      }
+      const map = getMap();
+      if (!map) {
+        removeFrameOverlay();
+        return;
+      }
+      preserveRenderedFrame();
+      map.once("render", removeFrameOverlay);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in 8eaebe52. The render cleanup listener is now registered only when frame preservation successfully created an overlay.

File: packages/map/src/map-resize.ts [hunk 4]
```diff
@@ -46,6 +46,81 @@ export function createMapResizeScheduler({
...
+    let backgroundColor = "";
+    while (backgroundElement && (!backgroundColor || backgroundColor === "rgba(0, 0, 0, 0)")) {
+      backgroundColor = window.getComputedStyle(backgroundElement).backgroundColor;
+      backgroundElement = backgroundElement.parentElement;
+    }
+    context.fillStyle = backgroundColor || "#fff";
+    context.fillRect(0, 0, overlay.width, overlay.height);
+    const x = (overlay.width - snapshot.width) / 2;
+    const y = (overlay.height - snapshot.height) / 2;
+    context.drawImage(snapshot, x, y);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Resolved by design in 8eaebe52 and the updated PR description. Per visual feedback, the frame is intentionally centered without stretching so the globe remains circular.

File: packages/map/src/map-resize.ts [hunk 5]
```diff
@@ -46,6 +46,81 @@ export function createMapResizeScheduler({
...
+    if (!canvas || typeof createImageBitmap !== "function") return;
+    void createImageBitmap(canvas)
+      .then((bitmap) => {
+        cachedFrame?.close();
+        cachedFrame = bitmap;
+      })
+      .catch(() => undefined);
+  };
+  snapshotMap?.on("idle", cacheRenderedFrame);
+  cacheRenderedFrame();
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in 8eaebe52. The idle bitmap cache was removed; preservation now synchronously captures the current canvas guaranteed by preserveDrawingBuffer.

File: packages/map/src/map-resize.ts [hunk 6]
```diff
@@ -162,6 +210,7 @@ export function createMapResizeScheduler({
     clearWindowResizeState();
     cancelFrame();
     cancelTimer();
+    preserveRenderedFrame();
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts:272]: Done in 8eaebe52. Starting a panel resize now removes the prior overlay and detaches its render listener before preserving a new frame; a regression test covers the interleaving.

File: packages/map/src/map-resize.ts [hunk 7]
```diff
@@ -46,6 +46,81 @@ export function createMapResizeScheduler({
...
+    overlay.width = Math.round(container.clientWidth * window.devicePixelRatio);
+    overlay.height = Math.round(container.clientHeight * window.devicePixelRatio);
+    const context = overlay.getContext("2d");
+    if (!context) return;
+    let backgroundElement: HTMLElement | null = container;
+    let backgroundColor = "";
+    while (backgroundElement && (!backgroundColor || backgroundColor === "rgba(0, 0, 0, 0)")) {
+      backgroundColor = window.getComputedStyle(backgroundElement).backgroundColor;
+      backgroundElement = backgroundElement.parentElement;
+    }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in 8eaebe52. The theme background is computed only when a new overlay is created and reused for subsequent resize callbacks.

File: tests/map-resize.test.ts [hunk 8]
```diff
@@ -212,10 +257,14 @@ describe("createMapResizeScheduler", () => {
       harness.flushFrames();
     }
     assert.equal(harness.resizeCalls(), 0, "the previous frame stays on screen during the drag");
+    assert.equal(harness.overlayCount(), 1, "a stretched frame covers the growing container");
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/map-resize.test.ts]: Done in 8eaebe52. The assertion now describes a changing container instead of incorrectly saying it grows.

File: packages/map/src/map-resize.ts [hunk 9]
```diff
@@ -46,6 +46,43 @@ export function createMapResizeScheduler({
...
+      width: "100%",
+      height: "100%",
+      pointerEvents: "none",
+    });
+    const context = overlay.getContext("2d");
+    if (!context) return;
+    context.drawImage(canvas, 0, 0);
+    canvas.insertAdjacentElement("afterend", overlay);
+    frameOverlay = overlay;
+    overlayMap = map;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in 8eaebe52. The decorative overlay is aria-hidden and explicitly sits above the four effects canvases and main map canvas.

File: packages/map/src/map-resize.ts [hunk 10]
```diff
@@ -46,6 +46,73 @@ export function createMapResizeScheduler({
...
+      let backgroundElement: HTMLElement | null = container;
+      while (
+        backgroundElement &&
+        (!overlayBackgroundColor || overlayBackgroundColor === "rgba(0, 0, 0, 0)")
+      ) {
+        overlayBackgroundColor = window.getComputedStyle(backgroundElement).backgroundColor;
+        backgroundElement = backgroundElement.parentElement;
+      }
+    }
+    context.fillStyle = overlayBackgroundColor || "#fff";
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in bdca2877. A fully transparent ancestor chain now falls back to an opaque white overlay background.

File: packages/map/src/map-resize.ts [hunk 11]
```diff
@@ -78,8 +145,18 @@ export function createMapResizeScheduler({
...
+        removeFrameOverlay();
+        return;
+      }
+      const map = getMap();
+      if (!map) {
+        removeFrameOverlay();
+        return;
+      }
+      preserveRenderedFrame();
+      if (frameOverlay) map.once("render", removeFrameOverlay);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in bdca2877. Each overlay now arms at most one MapLibre render cleanup listener.

File: packages/map/src/map-resize.ts [hunk 12]
```diff
@@ -162,6 +240,8 @@ export function createMapResizeScheduler({
     clearWindowResizeState();
     cancelFrame();
     cancelTimer();
+    removeFrameOverlay();
+    preserveRenderedFrame();
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts:272]: Done in bdca2877. ResizeObserver callbacks now refresh the overlay backing dimensions during panel drags, with a size assertion in the regression test.

File: packages/map/src/map-resize.ts [hunk 13]
```diff
@@ -46,6 +46,73 @@ export function createMapResizeScheduler({
...
+      while (
+        backgroundElement &&
+        (!overlayBackgroundColor || overlayBackgroundColor === "rgba(0, 0, 0, 0)")
+      ) {
+        overlayBackgroundColor = window.getComputedStyle(backgroundElement).backgroundColor;
+        backgroundElement = backgroundElement.parentElement;
+      }
+    }
+    context.fillStyle = overlayBackgroundColor || "#fff";
+    context.fillRect(0, 0, overlay.width, overlay.height);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in bdca2877. Transparent ancestor backgrounds are normalized to an opaque white fallback before filling the overlay.

File: packages/map/src/map-resize.ts [hunk 14]
```diff
@@ -46,6 +46,73 @@ export function createMapResizeScheduler({
...
+    const x = (overlay.width - snapshot.width) / 2;
+    const y = (overlay.height - snapshot.height) / 2;
+    context.drawImage(snapshot, x, y);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in 4342c04e. Snapshot capture DPR is retained and its destination dimensions scale by the current-to-capture DPR ratio.

File: packages/map/src/map-resize.ts [hunk 15]
```diff
@@ -78,8 +156,21 @@ export function createMapResizeScheduler({
...
+      if (!mapNeedsResize()) {
+        removeFrameOverlay();
+        return;
+      }
+      const map = getMap();
+      if (!map) {
+        removeFrameOverlay();
+        return;
+      }
+      preserveRenderedFrame();
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in 2c4d8380. The initial scheduler sizing pass now resizes without creating a snapshot overlay.

File: packages/map/src/map-resize.ts [hunk 16]
```diff
@@ -46,6 +46,84 @@ export function createMapResizeScheduler({
...
+    const canvasParent = canvas.parentElement;
+    if (!canvasParent || canvas.width === 0 || canvas.height === 0) return;
+
+    const snapshot = frameSnapshot ?? document.createElement("canvas");
+    if (!frameSnapshot) {
+      snapshot.width = canvas.width;
+      snapshot.height = canvas.height;
+      const snapshotContext = snapshot.getContext("2d");
+      if (!snapshotContext) return;
+      snapshotContext.drawImage(canvas, 0, 0);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in 2c4d8380. WebGL capture failures now skip preservation and allow the normal map.resize() path to proceed.

File: packages/map/src/map-resize.ts [hunk 17]
```diff
@@ -46,6 +46,93 @@ export function createMapResizeScheduler({
...
+        backgroundElement &&
+        (!overlayBackgroundColor || overlayBackgroundColor === "rgba(0, 0, 0, 0)")
+      ) {
+        overlayBackgroundColor = window.getComputedStyle(backgroundElement).backgroundColor;
+        backgroundElement = backgroundElement.parentElement;
+      }
+      if (!overlayBackgroundColor || overlayBackgroundColor === "rgba(0, 0, 0, 0)") {
+        overlayBackgroundColor = "#fff";
+      }
+    }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts:108]: Done in cce75337. Both the transparent keyword and Chromium rgba normalization are recognized.

File: packages/map/src/map-resize.ts [hunk 18]
```diff
@@ -46,6 +46,93 @@ export function createMapResizeScheduler({
...
+      position: "absolute",
+      inset: "0",
+      width: "100%",
+      height: "100%",
+      zIndex: "5",
+      pointerEvents: "none",
+    });
+    canvas.insertAdjacentElement("afterend", overlay);
+    frameOverlay = overlay;
+    overlayMap = map;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts:136]: Done in cce75337. A changed map instance now removes the old overlay/listener before capturing against the new map.

File: packages/map/src/map-resize.ts [hunk 19]
```diff
@@ -46,6 +46,84 @@ export function createMapResizeScheduler({
...
+
+    const snapshot = frameSnapshot ?? document.createElement("canvas");
+    if (!frameSnapshot) {
+      snapshot.width = canvas.width;
+      snapshot.height = canvas.height;
+      const snapshotContext = snapshot.getContext("2d");
+      if (!snapshotContext) return;
+      snapshotContext.drawImage(canvas, 0, 0);
+      frameSnapshot = snapshot;
+      frameSnapshotDevicePixelRatio = window.devicePixelRatio;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts]: Done in cce75337. Capture DPR is now derived from the canvas backing-to-CSS width ratio, with a DPR-only destination-width assertion.

File: packages/map/src/map-resize.ts [hunk 20]
```diff
@@ -46,6 +46,93 @@ export function createMapResizeScheduler({
...
+    if (frameOverlay) return;
+
+    overlay.className = "geolibre-map-resize-frame";
+    overlay.setAttribute("aria-hidden", "true");
+    Object.assign(overlay.style, {
+      position: "absolute",
+      inset: "0",
+      width: "100%",
+      height: "100%",
+      zIndex: "5",
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/map/src/map-resize.ts:131]: Confirmed visually in both themes: MapLibre controls remain visible because this z-index is confined to the canvas container stacking context. The overlay must remain above GeoLibre effects canvases 0–3 and the main canvas at 4.
